File: design/Stack.cpp
#include <cassert>
#include <string>
#include <optional>
#include <vector>

template <typename T>
class Stack {
private:
    std::vector<T> items = {};

public:
    Stack() = default;  // Constructor

    void push(const T& item) {
        items.push_back(item);
    }

    std::optional<T> pop() {
        if (items.empty()) return std::nullopt;

        T item = items.back();
        items.pop_back();
        return item;
    }

    std::optional<T> peek() {
        if (items.empty()) return std::nullopt;

        return items.back();   // top of the stack
    }

    size_t size() const {
        return items.size();
    }

    bool isEmpty() const {
        return items.empty();
    }
};

int main() {
    Stack<int> stack;

    assert(stack.isEmpty() == true);
    assert(stack.size() == 0);
    assert(!stack.pop().has_value());
    assert(!stack.peek().has_value());

    stack.push(1);
    assert(stack.isEmpty() == false);
    assert(stack.size() == 1);
    assert(stack.peek().value() == 1);

    stack.push(2);
    assert(stack.size() == 2);
    assert(stack.peek().value() == 2);

    assert(stack.pop().value() == 2);
    assert(stack.size() == 1);
    assert(stack.peek().value() == 1);

    assert(stack.pop().value() == 1);
    assert(stack.isEmpty() == true);
    assert(!stack.pop().has_value());
    assert(!stack.peek().has_value());

    stack.push(10);
    stack.push(20);
    stack.push(30);

    assert(stack.peek().value() == 30);
    assert(stack.pop().value() == 30);
    assert(stack.peek().value() == 20);
    assert(stack.size() == 2);

    Stack<std::string> s2;
    s2.push("hello");
    assert(s2.peek().value() == "hello");
    assert(s2.pop().value() == "hello");
    assert(s2.isEmpty() == true);

    Stack<double> s3;
    s3.push(3.14);
    assert(s3.pop().value() == 3.14);
    assert(s3.isEmpty() == true);

    return 0;
}

File: design/Stack.py
from typing import Any, List, Optional


class Stack:
    def __init__(self):
        self.items: List[Any] = []

    def push(self, item: Any):
        self.items.append(item)
    
    def pop(self) -> Optional[Any]:
        if self.items:
            return self.items.pop()
        return None
    
    def peek(self) -> Optional[Any]:
        if self.items:
            return self.items[-1]
        return None

    def size(self) -> int:
        return len(self.items)

    def is_empty(self) -> bool:
        return self.items == []


if __name__ == "__main__":
    stack = Stack()

    assert stack.is_empty() is True
    assert stack.size() == 0
    assert stack.pop() is None
    assert stack.peek() is None

    stack.push(1)
    assert stack.is_empty() is False
    assert stack.size() == 1
    assert stack.peek() == 1

    stack.push(2)
    assert stack.size() == 2
    assert stack.peek() == 2

    assert stack.pop() == 2
    assert stack.size() == 1
    assert stack.peek() == 1

    assert stack.pop() == 1
    assert stack.is_empty() is True
    assert stack.pop() is None
    assert stack.peek() is None

    stack.push(10)
    stack.push(20)
    stack.push(30)

    assert stack.peek() == 30
    assert stack.pop() == 30
    assert stack.peek() == 20
    assert stack.size() == 2

    stack.push("hello")
    assert stack.peek() == "hello"
    assert stack.pop() == "hello"

    stack.push(3.14)
    assert stack.pop() == 3.14
    assert stack.is_empty() is False

    stack.pop()
    stack.pop()
    assert stack.is_empty() is True
